### ISSUE TITLE
[RFC] Deprecate the `Slice` method of `DMatrix`.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#5556]
This is a proposal that we delegate data slicing to external libraries.

** Motivation

`DMatrix` is an internal data interface, used for all XGBoost algorithm.  The `DMatrix` object has a few different representation, based on which some more variants can be generated at runtime.  Currently we have 3 different base representation for `DMatrix`, namely a CSR like `SparsePage`, a Ellpack sparse matrix for storing GPU histogram index, and lastly external memory batched support for above two in-memory formats.  

The `Slice` method for `DMatrix` is mostly used for gridsearching where we need to split data for folds.  But the implementation is rather insufficient for many use cases as it supports only the `SparsePage` and without shuffling support, not to mention distributed environment support.  Also when used with ranking, it doesn't maintain the `group_ptr_`, which is delegated to a higher level Python wrapper script, which means the support status is not consistent across different language bindings.

** Moving forward

On the Python side, we have very good libraries of grid searching in `scikit-learn` and I believe it's not necessary for XGBoost to maintain a specialized implementation.  Also, we just added an inplace prediction function which can be extended with caching capability.  So performance is not of major concern.  We can easily delegate data slicing to external data inputs like `numpy` or `cudf` where slicing is implemented efficiently.

** Possible difficulty

Language bindings other than Python might have issue on supporting grid searching.  Currently `mkfold` is implemented on both R and Java.  While the `slice` method is implemented in R/JAVA/Scala .  Also I would very much like to keep capability with Julia/Go/Ruby bindings that are out of tree.  So the problem is we need to figure out what's canonical way of doing grid searching in all these bindings and try to integrate with them.

**Comment by RAMitchell** [comment]
+1 delegating slicing to other libraries will make our lives much easier if it can be achieved with reasonable performance.

**Comment by hcho3** [comment]
How does Dask ML deal with grid search and cross validation?

**Comment by trivialfis** [comment]
@hcho3 It's a problem that I want to resolve in next release, along with early stopping.  We need specialized implementation for dask grid searching, I'm hoping that we can implement something on dask-ml instead of XGBoost.

**Comment by hcho3** [comment]
@trivialfis That's good to hear. We'll need a performant method for handling cross validation. For example, I hope that we would not duplicate elements when we create a 5-fold split of training data.

**Comment by trivialfis** [comment]
Not sure if it's useful https://www.r-bloggers.com/r-setup-a-grid-search-for-xgboost/ .

**Comment by mayer79** [comment]
@trivialfis : I stumbled over this issue while writing a `split` method for `xgb.DMatrix` objects in R. 

I am curious about the status on `slice` and `xgb.cv`? Early stopping is a core element of model tuning. I don't think that standard ML packages like Scikit-Learn support a clean, *round-by-round evaluation* of early stopping during cross-validation With such packages, each cross-validation fold would get its "own" optimal number of boosting rounds instead of a "common" tree count (because their folds do not communicate with each other). Thus said, I would really miss `slice` and `xgb.cv` without good alternatives.

What could be interesting: We could add an optional argument `valid_ind` to `xgb.train()`. It would specify a vector of row indices of `data` used for validation. All other rows in `data` would be used for training. The necessary slicing would be applied internally, hopefully solving memory duplications. And getting rid of `slice`.

**Comment by trivialfis** [comment]
@mayer79 Thanks for the detailed explanation.  The issue is updated.  I recently toke a closer look into the cv implementation in xgboost and discussed related issues with @hcho3 .  The cv is likely to stay along with all the features, now I want to move the slicing into 3 party libraries from `DMatrix`.  Using slice method from general data libraries can help reducing both memory usage and complexity in `DMatrix`.  Sorry for the confusion, I should have taken more time before opening an issue.

**Comment by mayer79** [comment]
@trivialfis : Thanks a lot for clarification. Looking forward to the slice-free CV! Let me know if help is needed with the R Code.
